File: app/src/main/java/org/iggg/anaqram/MainActivity.java
package org.iggg.anaqram;

import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.SparseArray;
import android.view.Gravity;
import android.view.SurfaceView;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.vision.Detector;
import com.google.android.gms.vision.barcode.Barcode;

public class MainActivity extends AppCompatActivity {

    private TextView barcodeInfo;
    private CharBoxMapper charBoxMapper;
    private GameManager gameManager;
    private BarcodeDetectorManager barcodeDetectorManager;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        updateSetting();

        /* 下部のメニューボタンを描画 */
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });

        /* 中央部のカメラを描画 */
        SurfaceView cameraView = (SurfaceView) findViewById(R.id.camera_view);
        barcodeDetectorManager = new BarcodeDetectorManager(cameraView, this);

        barcodeInfo = (TextView) findViewById(R.id.code_info);

        barcodeDetectorManager.setProcessor(new Detector.Processor<Barcode>() {
            @Override
            public void release() { }

            @Override
            public void receiveDetections(Detector.Detections<Barcode> detections) {
                final SparseArray<Barcode> barcode = detections.getDetectedItems();
                if (barcode.size() != 0) {
                    barcodeInfo.post(new Runnable() {    // Use the post method of the TextView
                        public void run() {
                            String qrText = barcode.valueAt(0).displayValue;
                            barcodeInfo.setText(gameManager.displayChar(qrText));
                            charBoxMapper.updateChar();
                        }
                    });
                }
            }
        });

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            Intent intent = new Intent(this, SettingsActivity.class);
            startActivity(intent);
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        super.onResume();
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        if (!prefs.getString("answer","").equals(gameManager.getAnswer()))
            updateSetting();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        barcodeDetectorManager.release();
    }

    void updateSetting() {
        final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        gameManager = new GameManager(prefs.getString("answer",""));

        /* 上部のボタン文字列を描画 */
        LinearLayout buttonArea = (LinearLayout) findViewById(R.id.buttonArea);

        View.OnClickListener listener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
            charBoxMapper.swapCharBox((Button) v);
            String msg = gameManager.accept(charBoxMapper.getCurrentString());
            if (msg != null)
                toastMake(msg, 0, 0);

            }
        };

        charBoxMapper = new CharBoxMapper(this, gameManager.getCharBoxes(), listener);
        charBoxMapper.shuffle();
        charBoxMapper.updateChar();

        buttonArea.removeAllViews();
        for (Button button : charBoxMapper.getButtons())
            buttonArea.addView(button);
    }

    private void toastMake(String message, int x, int y) {
        TextView text = new TextView(this);
        text.setText(message);
        text.setTextSize(50);
        text.setTextColor(Color.WHITE);
        text.setBackgroundColor(ContextCompat.getColor(this, R.color.colorPrimary));

        Toast toast = new Toast(this);
        toast.setGravity(Gravity.CENTER, x, y);
        toast.setDuration(Toast.LENGTH_LONG);
        toast.setView(text);

        toast.show();
    }
}

File: app/src/main/java/org/iggg/anaqram/MainActivity.kt
package org.iggg.anaqram

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.Gravity
import android.view.SurfaceView
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast

import com.google.android.gms.vision.Detector
import com.google.android.gms.vision.barcode.Barcode

class MainActivity : AppCompatActivity() {
    data internal class Model(val gameManager: GameManager, val charBoxMapper: CharBoxMapper)
    private var model: Model? = null
    private var barcodeDetectorManager: BarcodeDetectorManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(findViewById(R.id.toolbar) as Toolbar)

        val startButton = findViewById(R.id.start_button) as Button
        val barcodeInfo = findViewById(R.id.code_info) as TextView

        startButton.setOnClickListener {
            model!!.run {
                if (gameManager.isRunning) {
                    gameManager.reset()
                    charBoxMapper.shuffle()
                    charBoxMapper.updateChar()
                    barcodeInfo.text = getString(R.string.please_start)
                    startButton.text = getString(R.string.start)
                } else {
                    gameManager.start()
                    barcodeInfo.text = getString(R.string.please_scan)
                    startButton.text = getString(R.string.reset)
                }
            }
        }

        updateSetting()

        /* 中央部のカメラを描画 */
        val cameraView = findViewById(R.id.camera_view) as SurfaceView
        barcodeDetectorManager = BarcodeDetectorManager(cameraView, this)
            .apply {
                setProcessor(object : Detector.Processor<Barcode> {
                    override fun release() = Unit
                    override fun receiveDetections(detections: Detector.Detections<Barcode>) =
                        model!!.run {
                            val barcode = detections.detectedItems
                            if (barcode.size() != 0 && gameManager.isRunning) {
                                barcodeInfo.post {
                                    val qrText = barcode.valueAt(0).displayValue
                                    barcodeInfo.text = gameManager.displayChar(qrText)
                                    charBoxMapper.updateChar()
                                }
                            }
                        }
                })
            }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId

        if (id == R.id.action_settings) {
            val intent = Intent(this, SettingsActivity::class.java)
            startActivity(intent)
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        model!!.let {
            if (prefs.getString("answer", "") != it.gameManager.answer) {
                it.gameManager.reset()
                updateSetting()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        barcodeDetectorManager?.release()
    }

    private fun updateSetting() {
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        val timerText = findViewById(R.id.timer) as TextView
        /* 上部のボタン文字列を描画 */
        val buttonArea =
            (findViewById(R.id.buttonArea) as LinearLayout).apply { removeAllViews() }

        val gameManager = GameManager(prefs.getString("answer", ""), timerText)
        val charBoxMapper = CharBoxMapper(this, gameManager.charBoxes, { str ->
            gameManager.accept(str)?.let { toastMake(it, 0, 0) }
        }).apply {
            shuffle()
            updateChar()
            buttons.forEach { button -> buttonArea.addView(button) }
        }
        model = Model(gameManager, charBoxMapper)
    }

    private fun toastMake(message: String?, x: Int, y: Int) =
        Toast(this)
            .apply { setGravity(Gravity.CENTER, x, y) }
            .apply { duration = Toast.LENGTH_LONG }
            .also { toast ->
                toast.view = TextView(this)
                    .also {
                        it.text = message
                        it.textSize = 50f
                        it.setTextColor(Color.WHITE)
                        it.setBackgroundColor(ContextCompat.getColor(this, R.color.colorPrimary))
                    }
            }.show()
}